### DISCUSSION TITLE
Consider adding @template to collections

### DISCUSSION THREAD

**Discussion by AAllport** [laravel/framework#34110] [Ideas] [answered]
This is a feature request to add `@template` to Collections

It's not uncommon to use tools such as PhpStan to ensure code quality within a Laravel project.
Furthermore, collections are commonly used as a return type, due to their superiority.

I propose adding the `@template` PhpDoc annotation to collections, to allow functions to annotate a return of a typed collection (granted, albeit at an ide/code-quality level)
Information about the tag can be found here: https://phpstan.org/blog/generics-in-php-using-phpdocs

Hopefully, adding this to such a commonly used framework will push support in PHPStorm https://youtrack.jetbrains.com/issue/WI-47158

**Comment by mfn** [comment]
phpstan isn't the only tool, is it compatible with other analyzers? e.g. psalm

(disclaimer: only phpstan user myself, but I can see laravel would only want things generically accepted; just a hunch)

**Reply by AAllport** [reply]
As far as I can tell, yes, it works: https://psalm.dev/docs/annotating_code/templated_annotations/

**Reply by b1rdex** [reply]
Yes! `@template` is a common syntax that is supported by the mentioned tools. Also, it's supported by Phan: https://github.com/phan/phan/wiki/Generic-Types

**Comment by ls-paul-julien-vauthier** [comment]
It looks like there was a previous attempt at introducing `@template` to collections that got rejected. See the discussion at: https://github.com/laravel/framework/pull/31486

**Reply by ondrejmirtes** [reply]
PhpStorm now understands PHPStan's generic syntax for better understanding of collections so this decision might be worthy reconsidering.

**Reply by AAllport** [reply]
Looks like there was quite a large changeset with that one.
I guess we could consider a slow and steady approach, (Like just seeing them into collections to gauge the suitability in the core).
And then gradually expand them, possibly looking into using better docs in QueryBuilder for returning a "typed" collection when running `->get()`

Cc: @GrahamCampbell, I know you were involved in that PR, any thoughts?

**Comment by AAllport** [comment, marked as answer]
Fixed in: #38538
